modules/core/src/test/java/org/apache/ignite/internal/processors/cache/distributed/dht/GridCachePartitionsStateValidationTest.java

```diff
@@ -323,4 +332,84 @@ public void waitUntilAllSingleMessagesAreSent() throws IgniteCheckedException {
             allSingleMessagesSent.get();
         }
     }
+
+    /**
+     *  Three tests that partitions state validation works correctly and show partition size always (GG-24471):
+     *
+     * Start three-nodes grid: first node (coordinator) with LogListener and two nodes without,
+     * arguments: cns - partition counters are inconsistent(boolean)
+     *           size - partition size are inconsistent(boolean)
+     */
+    private void startGridWithFirsLoggerNode(boolean cnt, boolean size) throws Exception {
+        testLogger = new ListeningTestLogger();
+
+        Collection<LogListener> lsnrs = new ArrayList<>();
+
+        if (cnt)
+            lsnrs.add(LogListener.matches("Partitions update counters are inconsistent for ").build());
+        if (size)
+            lsnrs.add(LogListener.matches("Partitions cache sizes are inconsistent for ").build());
+
+        lsnrs.forEach(testLogger::registerListener);
+
+        IgniteEx ignite = startGrid(0);
+            startGrid(1);
+            startGrid(2);
+            ignite.cluster().active(true);
+            awaitPartitionMapExchange();
+
+        // Populate cache to increment update counters.
+        for (int i = 0; i < 1000; i++)
+            ignite.cache(CACHE_NAME).put(i, i);
+
+        if (cnt){
+            // Modify update counter for some partition.
+            ignite.cachex(CACHE_NAME).context().topology().localPartitions().get(0).updateCounter(100500L);
+        }
+
+        if (size){
+            // Modify update size for some partition.
+            ignite.cachex(CACHE_NAME).context().topology().localPartitions().get(0).dataStore().clear(ignite.cachex(CACHE_NAME).context().cacheId());
+        }
+
+        // Trigger exchange.
+        startGrid(3);
+
+        awaitPartitionMapExchange();
+
+        // Nothing should happen (just log error message) and we're still able to put data to corrupted cache.
+        ignite.cache(CACHE_NAME).put(0, 0);
+
+        lsnrs.forEach(l->assertTrue(l.check()));
```

Comment:
I'd also add few checks in order to enusure that we don't log unexpected messages:
1. startGridWithFirsLoggerNode(true, false) -> "Partitions update counters are inconsistent for should be present" and "Partitions cache sizes are inconsistent for" should not be present;
2. startGridWithFirsLoggerNode(false, true) ->  "Partitions update counters are inconsistent for should not be present" and "Partitions cache sizes are inconsistent for" should be present;
3. startGridWithFirsLoggerNode(true, true) -> Both exception messages are expected;
4.  startGridWithFirsLoggerNode(false, false) -> No "Partitions update..." messages are expected;